src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,523 @@
+                root: earnerTokenRoot,
+                index: tokenLeafIndex,
+                proof: tokenProof,
+                leaf: tokenLeafHash
+            }),
+            "PaymentCoordinator._verifyTokenClaim: invalid token claim proof"
+        );
+    }
+
+    /**
+     * @notice verify inclusion of earner claim proof in the distribution root. This verifies
+     * the inclusion of the earner and earnerTokenRoot hash in the tree. The token claims are proven separately
+     * against the earnerTokenRoot hash (see _verifyTokenClaimProof). The earner leaf comprises of (earner, earnerTokenRoot)
+     * @param root distribution root that should be read from storage
+     * @param earnerLeafIndex index of the earner leaf
+     * @param earnerProof proof of the earners account root in the merkle tree
+     * @param earner address of earner account.
+     * @param earnerTokenRoot root hash of the token subtree for the earner.
+     */
+    function _verifyEarnerClaimProof(
+        bytes32 root,
+        uint32 earnerLeafIndex,
+        bytes calldata earnerProof,
+        address earner,
+        bytes32 earnerTokenRoot
+    ) internal view {
+        // Verify inclusion of earner leaf
+        bytes32 earnerLeafHash = calculateEarnerLeafHash(earner, earnerTokenRoot);
+        require(
+            Merkle.verifyInclusionSha256({
+                root: root,
+                index: earnerLeafIndex,
+                proof: earnerProof,
+                leaf: earnerLeafHash
+            }),
+            "PaymentCoordinator._verifyEarnerClaimProof: invalid earner claim proof"
+        );
+    }
+
+    /**
+     * @notice Sets the address of the entity that can claim payments on behalf of the earner
+     * @param earner The earner whose claimer is being set
+     * @param claimer The address of the entity that can claim payments on behalf of the earner
+     * @dev Only callable by the `earner`
+     */
+    function setClaimer(address earner, address claimer) external {
+        require(msg.sender == earner, "PaymentCoordinator.setClaimer: caller is not the earner");
+        address prevClaimer = claimerFor[earner];
+        claimerFor[earner] = claimer;
+        emit ClaimerForSet(earner, prevClaimer, claimer);
+    }
+
+    /// @notice return the hash of the earner's leaf
+    function calculateEarnerLeafHash(address earner, bytes32 earnerTokenRoot) public pure returns (bytes32) {
+        return sha256(abi.encode(earner, earnerTokenRoot));
+    }
+
+    /// @notice returns the hash of the earner's token leaf
+    function calculateTokenLeafHash(ClaimsTreeMerkleLeaf calldata leaf) public pure returns (bytes32) {
+        return sha256(abi.encode(leaf));
+    }
+
+    /// @notice returns 'true' if the claim would currently pass the check in `processClaims`
+    function checkClaim(PaymentMerkleClaim calldata claim) external view returns (bool) {
+        DistributionRoot memory root = distributionRoots[claim.rootIndex];
```

Comment:
Nit: This logic is essentially repeated twice - here and in `_verifyClaim`. Would prefer having it in one place. Recommend adding all of this functionality in a single function `_checkClaim`, which will also be called by `processClaim`. The `_processTokenClaim` function will then only need to transfer tokens, as the `_checkClaim` function will check earner and token roots.

Reply:
Done. `_processClaim` uses `_checkClaim` now. Originally didn't want to loop over tokenIndices again and do another storage read for claimerFor. I guess it shouldn't be too much more expensive though